### ISSUE TITLE
API Rate limit error while downloading `oc` causes Minishift to not start

### ISSUE THREAD

**Issue by mwcz** [minishift/minishift#1184]
I'm trying to get minishift set up on Fedora 26, and I see this each time I start it.

    $ ./minishift start
    Starting local OpenShift cluster using 'kvm' hypervisor...
    Error starting the cluster: Error attempting to download and cache oc: Cannot get the OpenShift release version v1.5.1: GET https://api.github.com/repos/openshift/origin/releases/tags/v1.5.1: 403 API rate limit exceeded for xxx.xxx.xxx.xxx.
     (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.); rate reset in 19m32.912254549s

It's nice that there's good news, but some instructions for how to switch to authenticated requests would be very helpful.  Other than trying again right after the rate limit resets, I'm not sure what else to do.  Other issues about rate limits suggest that it's not a big deal if it fails, but in my case this is the first run and it seems to be a necessary part of the first startup.

**Comment by gbraad** [comment]
This is actually an error related to the downloads allowed/requests to the
Github API. This should never happen and likely is an error due to Github

-- 

   Gerard Braad | http://gbraad.nl
   [ Doing Open Source Matters ]

**Comment by gbraad** [comment]
Please retry

**Comment by gbraad** [comment]
Can not reproduce...

/cc: @LalatenduMohanty this is either a glitch?

**Comment by LalatenduMohanty** [comment]
Right, @mwcz has hit the github rate limit for OC binary from openshift/origin project as Minishift.  The rate limit is quite high and users should not see this usually. The way to increase the rate limit is to use a OAuth2 github token or OAuth2 Key/Secret.  We use github token for our CI tests. 

@mwcz Please try again and you should be fine 

https://developer.github.com/v3/#increasing-the-unauthenticated-rate-limit-for-oauth-applications

**Comment by LalatenduMohanty** [comment]
I think we should this to troubleshooting guide and can point community there is they find the issue again. @mwcz @gbraad WDYT?

**Comment by mwcz** [comment]
Thanks @gbraad  and @LalatenduMohanty!

I tried several more times yesterday, and once today.  The rate limit has been exceeded each time.  It's not surprising; I'm at Red Hat Tower, a cradle of openshiftiness.  If you guys VPN in to RDU I bet you'll see the same thing.  I'll try it again from a different network and see if I can get past this.

Is there a way to ask minishift to skip the `oc` download?  In my case I already have `oc` installed.

**Comment by agajdosi** [comment]
@mwcz Hi, you can download the oc binary manually and put it into `~/.minishift/cache/oc/vx.x.x.x/`. There is just a catch that you need to realize the version of oc and create right named directory, currently it is `v1.5.1`.

Another way of going over this is to set environmental variable MINISHIFT_GITHUB_API_TOKEN. With token setup, your limit is higher and also is bind only to your persona.

**Comment by gbraad** [comment]
@mwcz let me know if you have success or not with one of the solutions. But yes, it might also happen on a customer's network of many developers decides to re-run installs often ;-). Adding it to the troubleshooting guide is not only helpful to you but also contributors when it fails on CI.

**Comment by gbraad** [comment]
@thatdocslady I re-organized the issue a little and changed the gist to 'adding an entry to the troubleshooting guide'. I think there is enough information in the ticket to get a basic idea. Just let me know if something isn't clear.

**Comment by mwcz** [comment]
It worked!  I VPNd into my home network and had no trouble downloading oc.  Thanks for the `MINISHIFT_GITHUB_API_TOKEN` tip as well, I missed that while trying to figure out how to activate auth'd API requests.  If you like, I could take a stab at updating the troubleshooting guide.  This quick test worked perfectly:

```
$ ./minishift update
Update failed: GET https://api.github.com/repos/minishift/minishift/releases/latest: 403 API rate limit exceeded for xxx.xxx.xxx.xxx. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.); rate reset in 40m3.144058799s

$ MINISHIFT_GITHUB_API_TOKEN="mytokenabcdef1234567890" ./minishift update
Nothing to update.
Already using the latest version: 1.3.1.
```

**Comment by gbraad** [comment]
Looking forward to the contribution :-D ... glad to see this working!

**Comment by mwcz** [comment]
I can't seem to get docs to build.  Might be the same as #552.

	$ eval $(minishift docker-env)

	$ make gen_docs
	glide install -v
	make: glide: Command not found
	make: *** [Makefile:62: vendor] Error 127

	$ make build_docs_container
	cd docs && docker build --build-arg uid=1000 -t minishift/docs .
	could not read CA certificate "/etc/docker/ca.pem": open /etc/docker/ca.pem: no such file or directory
	make: *** [Makefile:97: build_docs_container] Error 1

**Comment by gbraad** [comment]
as per https://github.com/minishift/minishift/issues/552#issuecomment-305160475 a workaround for F25 is known, but it should not affect F26...

**Comment by LalatenduMohanty** [comment]
> But yes, it might also happen on a customer's network if many developers decides to re-run installs often ;-)

@gbraad Do you mean with CDK? For upstream Minishift also it is very difficult to reproduce in non-RH environments as the rate limit is quite high. 
This will not happen with CDK because the OC binary as it wont fetch OC from github as it is embedded in the binary.

**Comment by mwcz** [comment]
Thanks @gbraad.  I don't know how to explain the `build_docs_container` issue.  I am on F26 and it isn't working.

I installed glide.  There's now a warning about "the" config file, and some import errors at the end.  I'm not familiar enough with golang to know where to go from here.

```
$ make gen_docs
glide install -v
[WARN]  The name listed in the config file (github.com/minishift/minishift) does not match the current location (.)
[INFO]  Downloading dependencies. Please wait...
[INFO]  --> Fetching github.com/cpuguy83/go-md2man.
[INFO]  --> Fetching github.com/docker/docker.
[INFO]  --> Fetching github.com/asaskevich/govalidator.
[INFO]  --> Fetching github.com/DATA-DOG/godog.
[INFO]  --> Fetching github.com/golang/protobuf.
[INFO]  --> Fetching github.com/blang/semver.
[INFO]  --> Fetching github.com/docker/go-units.
[INFO]  --> Fetching github.com/docker/machine.
[INFO]  --> Fetching github.com/fatih/color.
[INFO]  --> Fetching github.com/briandowns/spinner.
[INFO]  --> Fetching github.com/google/go-github.
[INFO]  --> Fetching github.com/magiconair/properties.
[INFO]  --> Fetching github.com/fsnotify/fsnotify.
[INFO]  --> Fetching github.com/google/go-querystring.
[INFO]  --> Fetching github.com/golang/glog.
[INFO]  --> Fetching github.com/hashicorp/hcl.
[INFO]  --> Fetching github.com/inconshreveable/go-update.
[INFO]  --> Fetching github.com/jteeuwen/go-bindata.
[INFO]  --> Fetching github.com/kardianos/osext.
[INFO]  --> Fetching github.com/inconshreveable/mousetrap.
[INFO]  --> Fetching github.com/mattn/go-colorable.
[INFO]  --> Fetching github.com/mattn/go-isatty.
[INFO]  --> Fetching github.com/mattn/go-runewidth.
[INFO]  --> Fetching github.com/mitchellh/mapstructure.
[INFO]  --> Fetching github.com/olekukonko/tablewriter.
[INFO]  --> Fetching github.com/pborman/uuid.
[INFO]  --> Fetching github.com/pelletier/go-buffruneio.
[INFO]  --> Fetching github.com/pelletier/go-toml.
[INFO]  --> Fetching github.com/pkg/browser.
[INFO]  --> Fetching github.com/pkg/errors.
[INFO]  --> Fetching github.com/russross/blackfriday.
[INFO]  --> Fetching github.com/samalba/dockerclient.
[INFO]  --> Fetching github.com/spf13/afero.
[INFO]  --> Fetching github.com/spf13/cast.
[INFO]  --> Fetching github.com/spf13/cobra.
[INFO]  --> Fetching github.com/spf13/jwalterweatherman.
[INFO]  --> Fetching github.com/spf13/pflag.
[INFO]  --> Fetching github.com/spf13/viper.
[INFO]  --> Fetching github.com/xeipuuv/gojsonschema.
[INFO]  --> Fetching golang.org/x/crypto.
[INFO]  --> Fetching golang.org/x/net.
[INFO]  --> Fetching golang.org/x/oauth2.
[INFO]  --> Fetching golang.org/x/sys.
[INFO]  --> Fetching golang.org/x/text.
[INFO]  --> Fetching google.golang.org/appengine.
[INFO]  --> Fetching gopkg.in/cheggaaa/pb.v1.
[INFO]  --> Fetching gopkg.in/yaml.v2.
[INFO]  Setting references.
[INFO]  --> Setting version for github.com/blang/semver to b38d23b8782a487059e8fc8773e9a5b228a77cb6.
[INFO]  --> Setting version for github.com/asaskevich/govalidator to 7b3beb6df3c42abd3509abfc3bcacc0fbfb7c877.
[INFO]  --> Setting version for github.com/DATA-DOG/godog to c3f5ce27930fff8efffaaf0acab7f24d5deecdbc.
[INFO]  --> Setting version for github.com/cpuguy83/go-md2man to bcc0a711c5e6bbe72c7cb13d81c7109b45267fd2.
[INFO]  --> Setting version for github.com/docker/go-units to 0bbddae09c5a5419a8c6dcdd7ff90da3d450393b.
[INFO]  --> Setting version for github.com/briandowns/spinner to 48dbb65d7bd5c74ab50d53d04c949f20e3d14944.
[INFO]  --> Setting version for github.com/docker/docker to a8a31eff10544860d2188dddabdee4d727545796.
[INFO]  --> Setting version for github.com/fatih/color to 62e9147c64a1ed519147b62a56a14e83e2be02c1.
[INFO]  --> Setting version for github.com/magiconair/properties to 51463bfca2576e06c62a8504b5c0f06d61312647.
[INFO]  --> Setting version for github.com/google/go-github to 30a21ee1a3839fb4a408efe331f226b73faac379.
[INFO]  --> Setting version for github.com/fsnotify/fsnotify to 4da3e2cfbabc9f751898f250b49f2439785783a1.
[INFO]  --> Setting version for github.com/inconshreveable/mousetrap to 76626ae9c91c4f2a10f34cad8ce83ea42c93bb75.
[INFO]  --> Setting version for github.com/golang/glog to 335da9dda11408a34b64344f82e9c03779b71673.
[INFO]  --> Setting version for github.com/google/go-querystring to 30f7a39f4a218feb5325f3aebc60c32a572a8274.
[INFO]  --> Setting version for github.com/inconshreveable/go-update to 8152e7eb6ccf8679a64582a66b78519688d156ad.
[INFO]  --> Setting version for github.com/docker/machine to 5b274558ea6ca822c06dd407a4e774a0105c3f60.
[INFO]  --> Setting version for github.com/golang/protobuf to 3c84672111d91bb5ac31719e112f9f7126a0e26e.
[INFO]  --> Setting version for github.com/hashicorp/hcl to 392dba7d905ed5d04a5794ba89f558b27e2ba1ca.
[INFO]  --> Setting version for github.com/kardianos/osext to ae77be60afb1dcacde03767a8c37337fad28ac14.
[INFO]  --> Setting version for github.com/jteeuwen/go-bindata to bbd0c6e271208dce66d8fda4bc536453cd27fc4a.
[INFO]  --> Setting version for github.com/mattn/go-isatty to 57fdcb988a5c543893cc61bce354a6e24ab70022.
[INFO]  --> Setting version for github.com/mattn/go-colorable to 5411d3eea5978e6cdc258b30de592b60df6aba96.
[INFO]  --> Setting version for github.com/mattn/go-runewidth to 97311d9f7767e3d6f422ea06661bc2c7a19e8a5d.
[INFO]  --> Setting version for github.com/mitchellh/mapstructure to db1efb556f84b25a0a13a04aad883943538ad2e0.
[INFO]  --> Setting version for github.com/pelletier/go-buffruneio to c37440a7cf42ac63b919c752ca73a85067e05992.
[INFO]  --> Setting version for github.com/pkg/errors to 645ef00459ed84a119197bfb8d8205042c6df63d.
[INFO]  --> Setting version for github.com/olekukonko/tablewriter to febf2d34b54a69ce7530036c7503b1c9fbfdf0bb.
[INFO]  --> Setting version for github.com/pkg/browser to c90ca0c84f15f81c982e32665bffd8d7aac8f097.
[INFO]  --> Setting version for github.com/spf13/afero to 9be650865eab0c12963d8753212f4f9c66cdcf12.
[INFO]  --> Setting version for github.com/russross/blackfriday to 0ba0f2b6ed7c475a92e4df8641825cb7a11d1fa3.
[INFO]  --> Setting version for github.com/pborman/uuid to 1b00554d822231195d1babd97ff4a781231955c9.
[INFO]  --> Setting version for github.com/spf13/cast to acbeb36b902d72a7a4c18e8f3241075e7ab763e4.
[INFO]  --> Setting version for github.com/pelletier/go-toml to 048765b4491bcff26505dfbb8a7b920133a19fd2.
[INFO]  --> Setting version for github.com/samalba/dockerclient to f661dd4754aa5c52da85d04b5871ee0e11f4b59c.
[INFO]  --> Setting version for github.com/spf13/jwalterweatherman to 0efa5202c04663c757d84f90f5219c1250baf94f.
[INFO]  --> Setting version for github.com/spf13/cobra to e606913c4ee45fec232e67e70105fb6c866b95d9.
[INFO]  --> Setting version for github.com/spf13/viper to 382f87b929b84ce13e9c8a375a4b217f224e6c65.
[INFO]  --> Setting version for github.com/xeipuuv/gojsonschema to 0c8571ac0ce161a5feb57375a9cdf148c98c0f70.
[INFO]  --> Setting version for github.com/spf13/pflag to e57e3eeb33f795204c1ca35f56c44f83227c6e66.
[INFO]  --> Setting version for golang.org/x/net to 4f2fc6c1e69d41baf187332ee08fbd2b296f21ed.
[INFO]  --> Setting version for golang.org/x/crypto to beef0f4390813b96e8e68fd78570396d0f4751fc.
[INFO]  --> Setting version for golang.org/x/text to 19e51611da83d6be54ddafce4a4af510cb3e9ea4.
[INFO]  --> Setting version for golang.org/x/sys to d9157a9621b69ad1d8d77a1933590c416593f24f.
[INFO]  --> Setting version for golang.org/x/oauth2 to 442624c9ec9243441e83b374a9e22ac549b5c51d.
[INFO]  --> Setting version for google.golang.org/appengine to 6a436539be38c296a8075a871cc536686b458371.
[INFO]  --> Setting version for gopkg.in/cheggaaa/pb.v1 to b6229822fa186496fcbf34111237e7a9693c6971.
[INFO]  --> Setting version for gopkg.in/yaml.v2 to cd8b52f8269e0feb286dfeef29f8fe4d5b397e0b.
[INFO]  Exporting resolved dependencies...
[INFO]  --> Exporting github.com/asaskevich/govalidator
[INFO]  --> Exporting github.com/DATA-DOG/godog
[INFO]  --> Exporting github.com/briandowns/spinner
[INFO]  --> Exporting github.com/docker/docker
[INFO]  --> Exporting github.com/blang/semver
[INFO]  --> Exporting github.com/docker/go-units
[INFO]  --> Exporting github.com/cpuguy83/go-md2man
[INFO]  --> Exporting github.com/docker/machine
[INFO]  --> Exporting github.com/fatih/color
[INFO]  --> Exporting github.com/magiconair/properties
[INFO]  --> Exporting github.com/fsnotify/fsnotify
[INFO]  --> Exporting github.com/jteeuwen/go-bindata
[INFO]  --> Exporting github.com/inconshreveable/go-update
[INFO]  --> Exporting github.com/hashicorp/hcl
[INFO]  --> Exporting github.com/mattn/go-colorable
[INFO]  --> Exporting github.com/google/go-querystring
[INFO]  --> Exporting github.com/golang/protobuf
[INFO]  --> Exporting github.com/google/go-github
[INFO]  --> Exporting github.com/golang/glog
[INFO]  --> Exporting github.com/mitchellh/mapstructure
[INFO]  --> Exporting github.com/mattn/go-isatty
[INFO]  --> Exporting github.com/inconshreveable/mousetrap
[INFO]  --> Exporting github.com/kardianos/osext
[INFO]  --> Exporting github.com/mattn/go-runewidth
[INFO]  --> Exporting github.com/pborman/uuid
[INFO]  --> Exporting github.com/olekukonko/tablewriter
[INFO]  --> Exporting github.com/pkg/browser
[INFO]  --> Exporting github.com/pelletier/go-buffruneio
[INFO]  --> Exporting github.com/spf13/cobra
[INFO]  --> Exporting github.com/samalba/dockerclient
[INFO]  --> Exporting github.com/spf13/afero
[INFO]  --> Exporting github.com/spf13/cast
[INFO]  --> Exporting github.com/pelletier/go-toml
[INFO]  --> Exporting github.com/spf13/jwalterweatherman
[INFO]  --> Exporting github.com/pkg/errors
[INFO]  --> Exporting github.com/xeipuuv/gojsonschema
[INFO]  --> Exporting github.com/spf13/viper
[INFO]  --> Exporting github.com/russross/blackfriday
[INFO]  --> Exporting github.com/spf13/pflag
[INFO]  --> Exporting golang.org/x/crypto
[INFO]  --> Exporting golang.org/x/net
[INFO]  --> Exporting golang.org/x/oauth2
[INFO]  --> Exporting golang.org/x/text
[INFO]  --> Exporting google.golang.org/appengine
[INFO]  --> Exporting golang.org/x/sys
[INFO]  --> Exporting gopkg.in/cheggaaa/pb.v1
[INFO]  --> Exporting gopkg.in/yaml.v2
[INFO]  Replacing existing vendor dependencies
[INFO]  Removing nested vendor and Godeps/_workspace directories...
[INFO]  Removing: /home/user/projects/minishift-git/vendor/github.com/cpuguy83/go-md2man/vendor
[INFO]  Removing: /home/user/projects/minishift-git/vendor/github.com/docker/docker/vendor
[INFO]  Removing: /home/user/projects/minishift-git/vendor/github.com/docker/machine/vendor
[INFO]  Removing: /home/user/projects/minishift-git/vendor/github.com/fatih/color/vendor
go get -u github.com/jteeuwen/go-bindata/...
package github.com/jteeuwen/go-bindata
        imports bufio: unrecognized import path "bufio" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports bytes: unrecognized import path "bytes" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports compress/gzip: unrecognized import path "compress/gzip" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports fmt: unrecognized import path "fmt" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports io: unrecognized import path "io" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports io/ioutil: unrecognized import path "io/ioutil" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports os: unrecognized import path "os" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports path/filepath: unrecognized import path "path/filepath" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports regexp: unrecognized import path "regexp" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports sort: unrecognized import path "sort" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports strings: unrecognized import path "strings" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports unicode: unrecognized import path "unicode" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata
        imports unicode/utf8: unrecognized import path "unicode/utf8" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata/go-bindata
        imports flag: unrecognized import path "flag" (import path does not begin with hostname)
package github.com/jteeuwen/go-bindata/go-bindata
        imports runtime: unrecognized import path "runtime" (import path does not begin with hostname)
make: *** [Makefile:84: /home/user/.go/bin/go-bindata] Error 1
```
I tried running it again for fun.  Less verbose, but same errors (oddly, the warning is gone).

```
$ make gen_docs
go get -u github.com/jteeuwen/go-bindata/...
package bufio: unrecognized import path "bufio" (import path does not begin with hostname)
package bytes: unrecognized import path "bytes" (import path does not begin with hostname)
package compress/gzip: unrecognized import path "compress/gzip" (import path does not begin with hostname)
package fmt: unrecognized import path "fmt" (import path does not begin with hostname)
package io: unrecognized import path "io" (import path does not begin with hostname)
package io/ioutil: unrecognized import path "io/ioutil" (import path does not begin with hostname)
package os: unrecognized import path "os" (import path does not begin with hostname)
package path/filepath: unrecognized import path "path/filepath" (import path does not begin with hostname)
package regexp: unrecognized import path "regexp" (import path does not begin with hostname)
package sort: unrecognized import path "sort" (import path does not begin with hostname)
package strings: unrecognized import path "strings" (import path does not begin with hostname)
package unicode: unrecognized import path "unicode" (import path does not begin with hostname)
package unicode/utf8: unrecognized import path "unicode/utf8" (import path does not begin with hostname)
package flag: unrecognized import path "flag" (import path does not begin with hostname)
package runtime: unrecognized import path "runtime" (import path does not begin with hostname)
make: *** [Makefile:84: /home/mclayton/.go/bin/go-bindata] Error 1

```

**Comment by gbraad** [comment]
it is best to do something like:

```
$ mkdir -p /workspace/src/github.com/minishift/ /workspace/bin /workspace/pkg
$ git clone git@github.com:minishift/minishift.git /workspace/src/github.com/minishift/minishift
$ export GOPATH /workspace
$ export PATH=$PATH:$GOPATH/bin
$ cd $GOPATH/src/github.com/minishift/minishift
$ make
```

It is a little tedious, but this is how go rolls...

I'll try to get the docs container built and see what your issue could be.

**Comment by gbraad** [comment]
Also, the issue with the `ca.pem` should not occur on Fedora 26. We reported it as #552 and it should have been fixed in newer versions... Did you run an upgrade from Fedora 25 to 26?

**Comment by thatdocslady** [comment]
@gbraad the error that @mwcz is reporting about the certificate was exactly what I was running into in F25 and the solution was described in: https://github.com/minishift/minishift/issues/552#issuecomment-287056337. @mwcz please try these steps, and keep us posted :)

General note about this issue: I didn't fedup yet to F26 but if the packages themselves haven't been updated it could be that we might still need this workaround (which means possibly reopening issue #552 or working from here).

**Comment by gbraad** [comment]
@praveenkumar informed me it is working for him with the distro provided
packafes on F26. So, new package installs would work... but since it is
config, it might have ended up as a .rpmnew instead.

**Comment by mwcz** [comment]
Yeah, I'm thinking that's what happened @gbraad , some stowaway file from when I upgraded (w/ dnf system upgrade plugin) from 25 to 26.  I'll try the GOPATH setup next week.  Do you think it would be worth adding that to the docs so non-golang users will be able to contribute to docs?

**Comment by gbraad** [comment]
I think @ThatDocsLady can contribute to this topic... as I am a developer
and I have an environment setup, I didn't really experience the situation
what actually needs to be there to make this work.

I go compile is needed AFAIR to generate documentation for the command and
parameters they take. But yes, you might be right that adding this is key.

**Comment by thatdocslady** [comment]
@gbraad I'm not a golang developer user but I do have the minishift setup on my machine because I need to work out of the repo as a part of my tech writing for the project. I didn't need to install Go or anything standalone other than minishift's dependencies, which Glide takes care of for me. 

As I wrote, I had the docker certificate problem when I tried to build the docs on F25 and what worked for me was to install the docker packages rather than try to reuse the minishift docker daemon. I opened an issue about it but the topic never ended up in the docs because the F25>26 upgrade was supposed to solve it so the devs closed the issue. 

This is everything I know about this issue, not sure what else I can contribute here :)
